### DISCUSSION TITLE
Can we make `:` optional when declaring functions?

### DISCUSSION THREAD

**Discussion by me2beats** [godotengine/godot-proposals#4699] [Scripting]
```gdscript
func f()
    return 0
```
It could still be possible to write `:` when you write lambdas and other oneliners

**Comment by Calinou** [comment]
I feel this reduces readability and adds choice paralysis (since there are now 2 syntaxes to choose from instead of one).

Programming languages generally stick to one of three forms of syntax and try to remain consistent throughout:

- Brace-based (C++, Java, …)
- Indentation-based (Python, GDScript, …)
- Keyword-based (Ruby, Lua, …)

Mixing and matching within the same language is often a bad idea, unless you can't avoid it.

**Reply by me2beats** [reply]
Well gdscript already has choices for example:
- write `;` at the end of a line
- use Lua or python styles when declaring a dictionary
- access a dictionary key in different ways, even using `.` (dict.key)
- write one liners
etc.
I think this is fine

**Comment by dalexeev** [comment]
Personally, I find it less readable, as if you accidentally forgot to end the line. This can probably also lead to parsing errors in some complex cases. Especially if this affects not only functions, but also other blocks (`if`, `for`, `while`, etc.). If not, it will introduce unnecessary inconsistency into the syntax.

I assume that the editor perceives pressing <kbd>Enter</kbd> after `:` as a sign to increase the indentation level on the next line. Without colon, the editor will need to check whether the expression is a block or a regular statement, which may not work for multiline blocks (for instance if the function has many parameters).

We prefer not to have multiple syntax options _doing the same thing_. For example, we have two dictionary styles, but the Lua style allows you to omit quotes for keys. We have two setter/getter styles, but the alternative style allows you to use a separate function like `setget` in 3.x. An alternative function declaration syntax would not bring any bonuses other than aesthetic ones. I think it's not worth it, it would only complicate parser, clutter documentation, etc.

Also, judging by the reactions and activity over almost 2 years, this proposal did not receive support. So I think it's reasonable to close this discussion. Thanks for the contribution nonetheless!
